man/hexproc.1
'\" t
.\"     Title: hexproc
.\"    Author: Filips Romāns
.\" Generator: Asciidoctor 2.0.10
.\"      Date: 2021-01-05
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "HEXPROC" "1" "2021-01-05" "\ \&" "\ \&"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
hexproc \- hexadecimal preprocessor for building binary files or hexdumps
.SH "SYNOPSIS"
.sp
\f(CR\fBhexproc\fP [\fIOPTION...\fP] [\fIFILE\fP]\fP
.SH "OPTIONS"
.sp
\fB\-v\fP
.RS 4
Prints the program version and exits
.RE
.sp
\fB\-V\fP
.RS 4
Prints advanced program configuration and exits
.RE
.sp
\fB\-h\fP
.RS 4
Prints a brief help message and exits
.RE
.sp
\fB\-b\fP
.RS 4
Outputs binary data instead of hexadecimal
.RE
.sp
\fB\-B\fP
.RS 4
Forces writing binary data even if output is a TTY
.RE
.sp
\fB\-c\fP
.RS 4
Enable colored output \- bytes resulting from a single
formatter or string will be marked with the same color;
not compatible with \fB\-b\fP or \fB\-B\fP
.RE
.sp
\fB\-C\fP
.RS 4
Forces colored output even if the output is not a TTY
.RE
.sp
\fB\-d\fP
.RS 4
Enter debug mode
.RE
.SH "DESCRIPTION"
.sp
Hexproc is a tool for building hex files. The input file
(or \f(CRstdin\fP, if no file given) is processed and then
written to \f(CRstdout\fP.
.sp
Hexproc is meant to be used as part of a pipeline and
therefore does not implement many features which are
provided by other processing tools.
.sp
It is recommended to use a text preprocessor (such as
\fBcpp\fP or \fBm4\fP) before feeding the input to \fBhexproc\fP as
\fBhexproc\fP does not provide macros or file inclusion.
.SH "SYNTAX"
.sp
Hexproc input consists of the following tokens:
.sp
\fBstrings\fP
.RS 4
Quoted ASCII text is converted to bytes and replaced
with hexadecimal octets. For example, the result of \fB"Hello"\fP
is \fB48 65 6c 6c 6f\fP.
.RE
.sp
\fBoctets\fP
.RS 4
Two hexadecimal digits will appear on the output as\-is.
.RE
.sp
\fBcomments\fP
.RS 4
End\-of\-line comments are created using \fB#\fP or \fB//\fP; Note that
comments in form of # \fINUMBER\fP "\fIfilename\fP" are recognized as
line markers, similar to how the C preprocessor uses them. This way,
you can get correct debug information after using \fB#include\fP.
Block comments start with \fB/*\fP and end with \fB*/\fP. They
cannot be nested.
.RE
.sp
\fBlabels\fP
.RS 4
The syntax \fINAME\fP \fB:\fP will assign the current byte offset
to the specified name. Labels may freely appear between tokens.
.RE
.sp
\fBlazy assignment\fP
.RS 4
The syntax \fINAME\fP \fB=\fP \fIEXPRESSION\fP will map the variable name to the
given expression without evaluating it. Whenever the name is
referenced, the expression will be evaluated. The expression must
not be recursive. If you want to add more tokens after an assignment,
terminate the expression with "\fB;\fP".
.RE
.sp
\fBimmediate assignment\fP
.RS 4
The syntax \fINAME\fP \fB:=\fP \fIEXPRESSION\fP will
evaluate the expression and assign the result to the variable. The
expression may reference the variable it\(cqs being assigned to. If
you want to add more tokens after an assignment, terminate the
expression with "\fB;\fP".
.RE
.sp
\fBformatters\fP
.RS 4
The syntax [\fIattr1\fP, \fIattr2\fP, \fI...\fP](\fIEXPRESSION\fP) is called a
formatter. Parentheses around \fIEXPRESSION\fP can be omitted if the
expression is a single number or variable name. The commas are optional.
Each \fIattr\fP can be one of the following:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
an integer between 0 and 8 inclusive, denoting the number of
bytes to use
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
the string \fBBE\fP or \fBLE\fP, which changes the representation of
the value to big or little endian, respectively
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
a type name, such as \fBint64\fP or \fBfloat\fP, which controls the
size and representation of the value. (See section \fBType Names\fP
for more information).
.RE
.RE
.sp
Leading and trailing whitespace is ignored.
.sp
Hexproc maps lines one\-to\-one so that line numbers
are preserved.
.SH "EXPRESSION SYNTAX"
.sp
Expressions are not standalone constructs, they are used in
assignments and formatters. Expressions support parentheses and the
following binary operators: \fB+\fP, \fB\-\fP, \fB*\fP, \fB/\fP, \fB^\fP
(exponentiation), \fB%\fP (remainder), \fB&\fP (binary and), \fB|\fP
(bitwise or) and \fB~\fP (bitwise xor). Unary minus and binary not are also
supported. You can use decimal, hexadecimal (prefix with \fB0x\fP) or octal
(prefix with \fB0\fP) number literals. The number literal format is
specified by the compiler used to build hexproc. You can also refer to
variables by their names. Operator precedence works just like in C.
.SH "SPECIAL VARIABLES"
.sp
A few variables have a special role in hexproc. These variable names
start with the prefix "\fBhexproc.\fP"
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\fBhexproc.endian\fP \- sets the default endianness of formatters
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\fBhexproc.major\fP \- version release number of hexproc ("\fB1\fP" in "\fBv1.2.3\fP")
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\fBhexproc.minor\fP \- minor version number of hexproc ("\fB2\fP" in "\fBv1.2.3\fP")
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\fBhexproc.patch\fP \- patch number of hexproc ("\fB3\fP" in "\fBv1.2.3\fP")
.RE
.SH "TYPE NAMES"
.sp
The following identifiers may be used as type names:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\fBint64\fP, \fBint32\fP, \fBint16\fP, \fBint8\fP \- fixed size types
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\fBlong\fP, \fBint\fP, \fBshort\fP, \fBbyte\fP \- synonymous with fixed size types
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\fBieee754_single\fP, \fBieee754_double\fP \- IEEE 754 single and double precision floating point types
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
\fBfloat\fP, \fBdouble\fP \- synonymous with above floating point types
.RE
.SH "DEBUGGER"
.sp
Hexproc comes with a built\-in debugger, which you can activate using the \fB\-d\fP option. The debugger supports the following commands:
.sp
\fBbreak\fP \fINUMBER\fP, \fBb\fP \fINUMBER\fP
.RS 4
Creates a breakpoint on given line number. When execution reaches a
line with a breakpoint, it will enter the debugger before evaluating the line
.RE
.sp
\fBresume\fP, \fBr\fP
.RS 4
Suspends the debugger and resumes execution.
.RE
.sp
\fBstep\fP, \fBs\fP
.RS 4
Executes next line and enters debugger again.
.RE
.sp
\fBvars\fP, \fBv\fP, \fBlist\fP, \fBl\fP
.RS 4
Displays the current state of all variables
.RE
.sp
\fBhelp\fP, \fBh\fP, \fB?\fP
.RS 4
Shows help on how to use the debugger
.RE
.sp
\fBeval\fP \fIEXPR\fP, \fBe\fP \fIEXPR\fP, \fB=\fP \fIEXPR\fP
.RS 4
Evaluate an expression
.RE
.SH "SEE ALSO"
.sp
xxd(1), cpp(1)
.SH "AUTHOR"
.sp
Filips Romāns

hexproc.c
#ifdef __TINYC__
void * __dso_handle __attribute((visibility("hidden"))) = &__dso_handle;
#endif

#include <stdlib.h>

#ifdef _POSIX_C_SOURCE
#include <unistd.h>
#elif defined(_WIN32)
#include <io.h>
// Since Windows is a special snowflake,
// here is a quick fix to get strtold to recognize hex numbers.
// Keep in mind that <stdlib.h> has to be included before this line
#define strtold(...) strtod(__VA_ARGS__)
#define __USE_MINGW_ANSI_STDIO 1
#include <stdio.h>
#else
#include <stdio.h>
int isatty(int);
int fileno(FILE*);
#endif

#include <assert.h>
#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <string.h>

#include "diagnostic.h"
#include "output.h"
#include "formatter.h"
#include "label.h"
#include "debugger.h"
#include "interpreter.h"

#ifndef HEXPROC_VERSION
#define HEXPROC_VERSION "-"
#endif

#ifndef HEXPROC_DATE
#define HEXPROC_DATE "?"
#endif

#ifndef HEXPROC_COMPILER
#define HEXPROC_COMPILER "?"
#endif

const size_t io_buffer_size = 8 * 1024;
const size_t default_linebuffer_size = 256;

static void print_usage(void) {
	fprintf(stderr,
"Usage: hexproc [OPTION...] [FILE]\n"
"  -v          Print program version and exit\n"
"  -V          Print program version and configuration and exit\n"
"  -h          Print this help message and exit\n"
"  -b          Output binary data\n"
"  -B          Force output binary data (even when output is a TTY)\n"
"  -c          Output colored text\n"
"  -C          Force output colored text (even when output is not a TTY)\n"
"  -d          Enable debugger\n"
"See the manual page hexproc(1) for more information\n"
	);
}

static void print_version(void) {
	printf("Hexproc %s\n", HEXPROC_VERSION);
}

static void print_configuration(void) {
#	ifdef HAVE_HP_FLOAT80
	#define HAVE_HP_FLOAT80_YESNO "yes"
#else
	#define HAVE_HP_FLOAT80_YESNO "no"
#endif
#	ifdef HAVE_HP_FLOAT128
	#define HAVE_HP_FLOAT128_YESNO "yes"
#else
	#define HAVE_HP_FLOAT128_YESNO "no"
#endif
#	ifdef HAVE_HP_INT128
	#define HAVE_HP_INT128_YESNO "yes"
#else
	#define HAVE_HP_INT128_YESNO "no"
#endif
	printf(
	"Version:         " HEXPROC_VERSION "\n"
	"Build timestamp: " HEXPROC_DATE "\n"
	"Compiler:        " HEXPROC_COMPILER "\n"
	"Have float80?  " HAVE_HP_FLOAT80_YESNO "\n"
	"Have float128? " HAVE_HP_FLOAT128_YESNO "\n"
	"Have int128?   " HAVE_HP_INT128_YESNO "\n"
	"Float expression type: " CALC_FLOAT_TYPENAME "\n"
	"Int expression type:   " CALC_INT_TYPENAME "\n"
	"Max expression call stack depth: %d\n"
	"Max number of expression tokens: %d\n"
	"=== Internal structure information ===\n"
	"Sizeof struct formatter: %d\n"
	"Sizeof struct label: %d\n"
	"Sizeof struct sourcemap: %d\n",
	(int)NAME_STACK_SIZE,
	(int)YARD_QUEUE_SIZE,
	(int)sizeof(struct formatter),
	(int)sizeof(struct label),
	(int)sizeof(struct sourcemap_entry)
	);
#undef HAVE_HP_FLOAT80_YESNO
#undef HAVE_HP_FLOAT128_YESNO
#undef HAVE_HP_INT128_YESNO
}

// special variables
static void add_builtin_variables(void) {
	set_constant_label(strdup("LE"), 0);
	set_constant_label(strdup("BE"), 1);
	set_constant_label(strdup("hexproc.endian"), 1);
	const char *version = HEXPROC_VERSION;
	long major = 0, minor = 0, patch = 0;
	char *end;
	version++; // skip the 'v' prefix
	major = strtol(version, &end, 0);
	if((version = end)[0] == '.')
		minor = strtol(++version, &end, 0);
	if((version = end)[0] == '.')
		patch = strtol(++version, &end, 0);

	set_constant_label(strdup("hexproc.major"), major);
	set_constant_label(strdup("hexproc.minor"), minor);
	set_constant_label(strdup("hexproc.patch"), patch);
}

void reset_terminal(void) {
	if(isatty(fileno(stdout)))
		fprintf(stdout, "\033[0m");
	if(isatty(fileno(stderr)))
		fprintf(stderr, "\033[0m");
}

int main(int argc, char **argv) {
	bool force_binary = false;
	bool force_color = false;

	opterr = 0; // disable 'getopt' error message
	int opt;
	while((opt = getopt(argc, argv, "vVhbBdcC")) != -1) {
		switch (opt) {
			case 'h':
				print_usage();
				return 0;
			case 'v':
				print_version();
				return 0;
			case 'V':
				print_configuration();
				return 0;
			case 'B':
				force_binary = true;
				output_mode = OUTPUT_BINARY;
				break;
			case 'b':
				output_mode = OUTPUT_BINARY;
				break;
			case 'd':
				debug_mode = true;
				break;
			case 'C':
				force_color = true;
				output_mode = OUTPUT_HEX_COLOR;
				break;
			case 'c':
				output_mode = OUTPUT_HEX_COLOR;
				break;
			default:
				fprintf(stderr, "Unknown option: %s\n", argv[optind]);
				print_usage();
				return EINVAL; // invalid argument
		}
	}

	if(isatty(fileno(stdout)) && (output_mode == OUTPUT_BINARY) && !force_binary) {
		fprintf(stderr, "Refusing to write binary data to console, use '-B' to override\n");
		return 1;
	}
	if(!isatty(fileno(stdout)) && (output_mode == OUTPUT_HEX_COLOR) && !force_color) {
		fprintf(stderr, "Refusing to write colored output to a non-tty, use '-C' to override\n");
		output_mode = OUTPUT_HEX;
		// not a fatal error, no need to exit
	}

	if(debug_mode)
		atexit(reset_terminal);

	if(optind >= argc) {
		// no file argument given
		current_input = stdin;
		current_file_name = "<stdin>";
	} else {
		current_input = fopen(argv[optind], "r");
		current_file_name = argv[optind];
		if(!current_input) {
			fprintf(stderr, "Couldn't open file \"%s\" (error %d)\n", argv[optind], (int) errno);
			return errno;
		}
	}

	char *line = malloc(default_linebuffer_size);
	size_t len = default_linebuffer_size;
	ssize_t nread;
	/* buffers for input/output streams, allocated with a single malloc call */
	char *shared_buffer = malloc(io_buffer_size * 2);


	if(!debug_mode && !isatty(fileno(current_input)))
		setvbuf(current_input, shared_buffer, _IOFBF, io_buffer_size);

	FILE *output = stdout;
	if(!debug_mode && !isatty(fileno(output)))
		setvbuf(output, shared_buffer + io_buffer_size, _IOFBF, io_buffer_size);

	add_builtin_variables();

	if(debug_mode && isatty(fileno(stdin)))
		if(signal(SIGINT, enter_debugger_async) == SIG_ERR)
			fprintf(stderr, "Internal error: could not setup signal handler for SIGINT, errno = %d\n", errno);

	if(debug_mode && isatty(fileno(stdin)))
		enter_debugger();

	struct bytequeue buffer = make_bytequeue();

	while((nread = getline(&line, &len, current_input)) != -1 || !feof(current_input)) {
		clearerr(current_input);
		line_number++;
		process_line(line, &buffer);
	}

	bytequeue_rewind(&buffer);

	line_number = 1;
	offset = 0;

	for(int byte; (byte = bytequeue_get(&buffer)) != EOF;)
		output_byte(byte, output);

	finalize_output(output);

	fflush(output);
	if(output != stdout)
		fclose(output);
	if(current_input != stdin)
		fclose(current_input);

	cleanup_formatters();
	cleanup_labels();
	cleanup_breakpoints();
	cleanup_sourcemap();

	free_bytequeue(buffer);

	OPTIONAL_FREE(line);
	OPTIONAL_FREE(shared_buffer);
	return 0;
}
